### ISSUE TITLE
WSL2: Seek of directory entry by lseek does not work on v9fs

### ISSUE THREAD

**Issue by ngyuki** [microsoft/WSL#5074]
**Build number: Microsoft Windows [Version 10.0.19603.1000]**

Seek of directory entry by lseek does not work on v9fs.

**Code: a.c**

```c
#define _GNU_SOURCE
#include <dirent.h>
#include <fcntl.h>
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <sys/syscall.h>

struct linux_dirent {
    long           d_ino;
    off_t          d_off;
    unsigned short d_reclen;
    char           d_name[];
};

#define BUF_SIZE 1024

int main()
{
    int fd, nread;
    char buf[BUF_SIZE];
    struct linux_dirent *d;
    off_t off;
    int bpos;

    fd = open("./xxx/", O_RDONLY | O_DIRECTORY);
    syscall(SYS_getdents, fd, buf, BUF_SIZE);
    off = lseek(fd, 0, SEEK_SET);
    if (off < 0) {
        return 1;
    }

    while ((nread = syscall(SYS_getdents, fd, buf, BUF_SIZE)) > 0) {
        for (bpos = 0; bpos < nread;) {
            d = (struct linux_dirent *)(buf + bpos);
            printf("%s\n", d->d_name);
            bpos += d->d_reclen;
        }
    }
    return 0;
}
```

**Prepare directory and running**

```sh
mkdir ./xxx/
seq -w 0000001 0000040 | xargs -i touch ./xxx/{}.txt
ls xxx/ | wc -l
#=> 40

gcc a.c
./a.out | sort
```

**Actual output**

```
0000031.txt
0000032.txt
0000033.txt
0000034.txt
0000035.txt
0000036.txt
0000037.txt
0000038.txt
0000039.txt
0000040.txt
```

**Expected output**

```
.
..
0000001.txt
0000002.txt
0000003.txt
0000004.txt
0000005.txt
0000006.txt
0000007.txt
0000008.txt
0000009.txt
0000010.txt
0000011.txt
0000012.txt
0000013.txt
0000014.txt
0000015.txt
0000016.txt
0000017.txt
0000018.txt
0000019.txt
0000020.txt
0000021.txt
0000022.txt
0000023.txt
0000024.txt
0000025.txt
0000026.txt
0000027.txt
0000028.txt
0000029.txt
0000030.txt
0000031.txt
0000032.txt
0000033.txt
0000034.txt
0000035.txt
0000036.txt
0000037.txt
0000038.txt
0000039.txt
0000040.txt
```

On ext4 we get the expected result.

**Strace**

```sh
execve("./a.out", ["./a.out"], 0x7ffdc9af5ce0 /* 20 vars */) = 0
brk(NULL)                               = 0x556e019c6000
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
access("/etc/ld.so.preload", R_OK)      = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=25456, ...}) = 0
mmap(NULL, 25456, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7fe00ba2c000
close(3)                                = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\260\34\2\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0755, st_size=2030544, ...}) = 0
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fe00ba2a000
mmap(NULL, 4131552, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7fe00b41b000
mprotect(0x7fe00b602000, 2097152, PROT_NONE) = 0
mmap(0x7fe00b802000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1e7000) = 0x7fe00b802000
mmap(0x7fe00b808000, 15072, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7fe00b808000
close(3)                                = 0
arch_prctl(ARCH_SET_FS, 0x7fe00ba2b4c0) = 0
mprotect(0x7fe00b802000, 16384, PROT_READ) = 0
mprotect(0x556e0166a000, 4096, PROT_READ) = 0
mprotect(0x7fe00ba33000, 4096, PROT_READ) = 0
munmap(0x7fe00ba2c000, 25456)           = 0
openat(AT_FDCWD, "./xxx/", O_RDONLY|O_DIRECTORY) = 3
getdents(3, /* 32 entries */, 1024)     = 1008
lseek(3, 0, SEEK_SET)                   = 0
getdents(3, /* 10 entries */, 1024)     = 320
fstat(1, {st_mode=S_IFCHR|0620, st_rdev=makedev(136, 0), ...}) = 0
brk(NULL)                               = 0x556e019c6000
brk(0x556e019e7000)                     = 0x556e019e7000
write(1, "0000031.txt\n", 120000031.txt
)           = 12
write(1, "0000032.txt\n", 120000032.txt
)           = 12
write(1, "0000033.txt\n", 120000033.txt
)           = 12
write(1, "0000034.txt\n", 120000034.txt
)           = 12
write(1, "0000035.txt\n", 120000035.txt
)           = 12
write(1, "0000036.txt\n", 120000036.txt
)           = 12
write(1, "0000037.txt\n", 120000037.txt
)           = 12
write(1, "0000038.txt\n", 120000038.txt
)           = 12
write(1, "0000039.txt\n", 120000039.txt
)           = 12
write(1, "0000040.txt\n", 120000040.txt
)           = 12
getdents(3, /* 0 entries */, 1024)      = 0
exit_group(0)                           = ?
+++ exited with 0 +++
```

**Comment by ngyuki** [comment]
The following code reproduces the same problem.

```c
#include <stdio.h>
#include <dirent.h>

int main() {
    DIR *dh;
    struct dirent *de;

    dh = opendir("./xxx/");
    readdir(dh);

    rewinddir(dh); // does not work

    while (de = readdir(dh)) {
        printf("%s\n", de->d_name);
    }

    return 0;
}
```

```sh
seq -w 0000001 0001030 | xargs -i touch ./xxx/{}.txt

ls xxx/ | wc -l
#=> 1030

gcc a.c
./a.out | sort
#=> 0001023.txt
#=> 0001024.txt
#=> 0001025.txt
#=> 0001026.txt
#=> 0001027.txt
#=> 0001028.txt
#=> 0001029.txt
#=> 0001030.txt
```

**Comment by therealkenc** [comment]
Nice post. Confirmed WSL2 19603.

As a work-around, I confirmed cifs is okay. You'll have to do without metadata though, natch.

![image](https://user-images.githubusercontent.com/7049700/79040093-fe4f2080-7b9a-11ea-9397-cd929368417a.png)

Something like:
```
$ sudo apt install cifs-utils
$ mkdir ./mnt
$ sudo mount -t cifs -o username=you //xxx.xxx.xxx.xxx/Users/you ./mnt
```

**Comment by themizzi** [comment]
Curious why this is closed since it does not seem to have been repaired. PHP maintainers believe this is the cause of the following issue: https://github.com/microsoft/WSL/issues/6081 which affects Docker for Desktop and makes this a chore. Is there any appetite to fix this with v9fs? I do not see a bug in the v9fs bug tracker regarding this.

https://bugs.php.net/bug.php?id=80227

**Comment by Zenexer** [comment]
Any update on this?  This breaks popular PHP applications that are likely to be run in WSL2 for development purposes, including XenForo.

**Comment by SeniorSupportEngineer** [comment]
I got problems with phpunit that does not discover all test files in a large directory. Find it strange that this still isn't fixed.

**Comment by VeselyJan92** [comment]
I there any temp workaround?

**Comment by VeselyJan92** [comment]
Suggestion: If wouldn't mind using VSCode it's actually better fully switch development into WSL and use VSCode as frontend editor more on that [here](https://code.visualstudio.com/docs/remote/wsl). There is no overhead between translating different file systems and works like a charm and there is no speed/ram/cpu overhead.  I am running a PHP docker app with its code sticking out for development.

**Comment by stevennight** [comment]
Any bro has been resolved?
I try to run the swoft project on docker on Windows, but some file can not read when loading.

**Comment by SailorMax** [comment]
Same problem with symfony/finder class, which uses php's RecursiveDirectoryIterator class.

**Comment by dacgray** [comment]
Same problem when using squizlabs/php_codesniffer

**Comment by ReWiG** [comment]
Confirming the same issue with phpunit/phpunit and zircote/swagger-php
Windows 10 Pro 21H2

**Comment by Zenexer** [comment]
Could we add a bug label to this?

**Comment by Zenexer** [comment]
@benhillis, could you put a bug label on this so it doesn't get lost?  This is a pretty serious issue that prevents quite a bit of software from working within WSL2.

**Comment by Ador-able** [comment]
I have encountered the same problem. Has the error from 3 years ago not been fixed yet

**Comment by NathanBaulch** [comment]
Still preventing me from launching docker containers with Windows FS volume mounts 😞, it's been years now.

**Comment by selfer-work** [comment]
I needed to set up my project for emergencies on a laptop. However this bug makes Docker Desktop on Windows 11 Home edition unusable when combining it with PHPStorm. Simple solution on Windows 11 Pro edition is to switch to Hyper-V. But no, for some reason you can't do that in Home edition. So no point in using Windows for me I guess.

**Comment by SeniorSupportEngineer** [comment]
@selfer-work That is why I switch to the Ubuntu app and now I'am using projector, phpstorm runs in projector in ubuntu, the client in windows uses the weblink. And yes it is not ideal. Works better than the windows for docker.

**Comment by Mattie112** [comment]
(a quick "fix" is to put all source files "in" WSL, then open that directory in your PhpStorm (or whatever), the WSL files should be available at \\wsl.local if I am not mistaking)

**Comment by SailorMax** [comment]
"dev drive" also can be used as "quick fix" for this problem: https://learn.microsoft.com/en-us/windows/dev-drive/

**Comment by selfer-one** [comment]
Great behavior. Just ignore the issues without reply. FU.

**Comment by Zenexer** [comment]
I've requested that this be reopened: https://github.com/microsoft/WSL/issues/12416

**Comment by jurchiks** [comment]
> "dev drive" also can be used as "quick fix" for this problem: https://learn.microsoft.com/en-us/windows/dev-drive/

I _am_ using Dev Drive, but I am still having this issue.
I cloned the code while in WSL, and I ran `docker compose up -d` while in WSL.
This is insane, almost 5 years of this issue and not a single official comment!

**Comment by Zenexer** [comment]
@OneBlue or @benhillis, could I bother you to reopen this and add the "bug" label?  This continues to prevent WSL2 users from working with PHP--and is probably the underlying cause of other sporadic bugs.

**Comment by jurchiks** [comment]
Yesterday I found out that apparently DevDrive is actually trash.

My DevDrive was using ReFS (not sure if any other FS is available), and I had cloned my source code into that DevDrive, which was mounted at `/mnt/f/`. Initially I had cloned the code from within Windows, then I tried cloning from within WSL, but that didn't change anything at all.

My issue was: running Doctrine proxy generation command failed to find any except 3-4 Entity files (last 3-4 in the corresponding directory for whatever reason),
and as a result running Doctrine db-diff command resulted in it generating a migration that deleted all tables except the 3-4 that it could find.

How I found out the fix:
when I was opening my project in PHPStorm, Windows was giving me a notification that "I/O operations might be slow when using Git for Windows". I googled that message and found this issue: https://github.com/microsoft/WSL/issues/9555
Which mentioned:
> Is this using a folder from the Windows file system or is it in the default WSL system?

So I thought - hmm, perhaps Dev Drive isn't actually a WSL file system?
I then downloaded Ubuntu 24.04 LTS and set my code up in `/var/www/`, and, [while I had some issues during the process](https://github.com/microsoft/WSL/issues/12781), the end result works well.

Now, the only potential issue is this: https://youtrack.jetbrains.com/issue/IJPL-2799/Running-it-in-Windows-Subsystem-for-Linux-External-file-changes-sync-may-be-slow-File-watcher-gave-up-to-operate

I will also add this just for extra helpful thing: https://superuser.com/questions/1183176/can-i-share-my-ssh-keys-between-wsl-and-windows
I have my SSH keys in my Windows user's home directory, and I wouldn't want them in WSL only.

**Comment by ReWiG** [comment]
This problem has never been solved.

**Comment by selfer-one** [comment]
> This problem has never been solved.

And it never will be. I switched to linux and didn't bother with windows anymore. Just switch, it's easier.

**Comment by denkfabrik-li** [comment]
This is still reproducible in August 2026 — WSL 2.6.3, Windows 11 Pro build 26200, via Docker Desktop (WSL2 backend) with a directory bind-mounted from `C:\` into a Linux container (`stat -f` inside the container reports `v9fs`). Since the issue is closed as completed, here is a concrete, fully diagnosed data point in case it helps triage a reopen or a new report.

**Symptom.** Any reader that rewinds a partially read directory handle silently loses the first read buffer of entries. PHP's SPL directory iterators (`FilesystemIterator`, `RecursiveDirectoryIterator`, and everything built on them, e.g. Symfony Finder) do exactly that: the constructor reads the first buffer of dirents, then `foreach` calls `rewind()` — the seek back to offset 0 is ignored, iteration resumes after the buffered chunk, and no error is raised anywhere. Plain `readdir()`, `scandir()`, `glob()`, `ls` and `find` never rewind mid-stream and are unaffected.

**Measurements** (directory of migration-style files with ~69-char names, counted from inside the container, August 2026):

| Reader | `php:8.3-cli-alpine` (musl), 200 files | `php:8.3-cli` (glibc), 500 files |
| --- | --- | --- |
| `scandir()` / `glob()` | 200 | 500 |
| `FilesystemIterator` | **179** | **157** |
| plain `readdir()` loop | 200 | 500 |

The dropped count is exactly one libc directory read buffer: musl reads directories in 2 KiB chunks (≈ 21 of these dirents), glibc in 32 KiB chunks (≈ 343). So on Alpine-based images the loss already starts at ~21 files per directory; on glibc at a few hundred. The same 200/500-file directory listed via a volume from the container's own ext4/overlay filesystem is always complete.

**Real-world impact.** PHPStan/Larastan scans `database/migrations` of a Laravel app with `RecursiveDirectoryIterator`. Over the 9p mount (Alpine PHP container) it saw 43 of 70 migration files and consequently reported 163 phantom "property does not exist on model" errors. The identical tree, byte-for-byte copied to ext4 inside the same container, analyses with 0 errors, matching GitHub-hosted CI. Nothing in the toolchain hints at the filesystem: every direct lookup of a "missing" file succeeds, only buffered iteration misses them. This matches php.net bugs #80056 / #80227 (both closed "not a PHP bug", pointing at this issue). Downstream write-up with the fix on the PHP side: larastan/larastan#2538.

This class of bug is nasty precisely because it is silent and partial — build tools, autoloader generators, test discovery and static analysers just see fewer files and carry on. A fix at the 9p layer (honoring directory seeks after partial reads, or invalidating the client-side buffer on rewind) would resolve a whole family of "works on ext4, mysteriously broken through Docker Desktop / WSL2 mounts" reports, e.g. microsoft/WSL#6081
